upgrade: Rework router migration

rhafer opened the pull request:
This is still very much work in progress. Especially the migration for subsequent nodes is largely untested. Just to give an overview where things are heading. Also there is a small conflict with compute node work in #875 (but that should be easy to solve)

rhafer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(5) S51P4 Solve network traffic transition during upgrade of controller nodes](https://trello.com/c/QhV7YPbo/185-5-s51p4-solve-network-traffic-transition-during-upgrade-of-controller-nodes)

abel-navarro commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -407,24 +410,24 @@ def upgrade_drbd_cluster(cluster)
         return false unless upgrade_first_cluster_node drbd_slave, drbd_master
 
-        return false unless upgrade_next_cluster_node drbd_master
+        return false unless upgrade_next_cluster_node drbd_master, drbd_slave
```
inside `upgrade_next_cluster_node` we [already call](https://github.com/crowbar/crowbar-core/pull/935/files#diff-33ff816d38c460e2a65a4918bf6f94d0R377) `upgrade_first_cluster_node` (line above).

rhafer replied:
@abel-navarro Not sure what you mean by that. Please elaborate.

abel-navarro replied:
@rhafer sorry my mistake. I didn't expand the sections on the PR diff and mistakenly assigned a call to the wrong function. Alles klar.

abel-navarro commented:
actually the >2 node cluster work was already addressed here: #779

rhafer commented on chef/cookbooks/crowbar/templates/default/crowbar-router-migration.sh.erb:
```diff
@@ -25,21 +27,6 @@ set +x
-    ssh $old_host ip netns delete qrouter-$router
-
-done
+/usr/bin/neutron-ha-tool --l3-agent-evacuate $hostname --wait-for-router
```
Yeah, that's one reason why I added the WIP label.

jsuchome commented on crowbar_framework/app/models/node_object.rb:
```diff
@@ -1123,6 +1123,16 @@ def wait_for_script_to_finish(script, seconds, args = [])
+  def delete_script_exit_files(script)
+    base = "/var/lib/crowbar/upgrade/" + File.basename(script, ".sh")
+    ok_file = base + "-ok"
+    failed_file = base + "-failed"
```
It would probably make sense to have a method that would return `[ ok_file, failed_file ]` for given script. We're using this mechanism already on 3 places in this model...

rhafer replied:
Hm, true. Are you ok if I address that in a separate PR?

jsuchome replied:
Yes

jsuchome commented:
Would it be possible to squash the last commit to the commit which broke the testsuite? 
Or is it rather fixing the tests that were broken by design and not failing?

jsuchome commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -445,6 +451,28 @@ def delete_pacemaker_resources(node_name)
+      # Evacuate all routers away from the specified network node to other
+      # available network nodes. The evacuation procedure it started on the
+      # specified controller node
+      def evacuate_network_node(controller, network_node, delete_namespaces = false)
```
So the `controller` node is the target node for the routers, right? Wouldn't it be better name?
(Feel free to discard this comment, if network experts think it is better this way, than it likely is)

rhafer replied:
The controller node is actually not the target. Well, it might be some cases (the standard single-two-node-cluster setup). But in most other cases it isn't.

jsuchome replied:
Ah, ok

jsuchome commented on chef/cookbooks/crowbar/templates/default/crowbar-router-migration.sh.erb:
```diff
@@ -10,36 +10,23 @@ UPGRADEDIR=/var/lib/crowbar/upgrade
-    ssh $old_host ip netns delete qrouter-$router
-
-done
+/usr/bin/neutron-ha-tool --l3-agent-evacuate $hostname --wait-for-router $delete_ns
```
Could this `neutron-ha-tool` invocation fail? 

Currently the only possible failure of this whole script is the timeout...

rhafer replied:
Fixed

jsuchome commented:
So, I've tested this manually, and noticed only few packet drops during the first invocation of the router migration script. Here's the relevant node-upgrade.log from the first node (edited for better readability): http://pastebin.nue.suse.com/18778/src

rhafer commented:
> Would it be possible to squash the last commit to the commit which broke the testsuite?

Sure.

> Or is it rather fixing the tests that were broken by design and not failing?

It's a mixture of both actually. But I guess it makes sense to squash it with the other commit.

jsuchome commented on chef/cookbooks/crowbar/templates/default/crowbar-router-migration.sh.erb:
```diff
@@ -10,36 +10,30 @@ UPGRADEDIR=/var/lib/crowbar/upgrade
+/usr/bin/neutron-ha-tool --l3-agent-evacuate $hostname --wait-for-router $delete_ns
+ret=$?
+if [ $ret != 0 ] ; then
+    echo "Failed to evacuate l3 agent on host: %hostname"
```
`$hostname`

rhafer replied:
fixed

matelakat commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -451,6 +457,28 @@ def delete_pacemaker_resources(node_name)
+        args = [network_node]
+        args << "delete-ns" if delete_namespaces
+        controller.wait_for_script_to_finish(
+          "/usr/sbin/crowbar-router-migration.sh", 600, args
```
This hardwired `600` looks strange, although I understand that this was the timeout originally for the other script. Given that we are using `--wait-for-router`, execution time can increase hugely. I am not sure we need to address it here, but at some point this hardwired value could be a real pain. Best would be to make the timeout's value depending on the number of routers. This is something that we might want to put inside neutron-ha-tool

rhafer replied:
Yeah, I am not really happy with that either. And 10 minutes is clearly not enough for a somewhat bigger setup.
We will need to rethink a few other timeouts here as well.

matelakat commented on crowbar_framework/spec/models/api/upgrade_spec.rb:
```diff
@@ -379,10 +394,6 @@
       allow_any_instance_of(Api::Node).to receive(:router_migration).and_return(true)
```
Do we need this line? Is it referring to the same method that we just removed?

rhafer replied:
Hm, right. I overlooked that.

matelakat commented on chef/cookbooks/crowbar/templates/default/crowbar-router-migration.sh.erb:
```diff
@@ -10,36 +10,30 @@ UPGRADEDIR=/var/lib/crowbar/upgrade
 
 mkdir -p $UPGRADEDIR
 
-if [[ -f $UPGRADEDIR/crowbar-router-migration-ok ]] ; then
```
Why are we removing this part? If I understand it correcly, if the migration is successful, crowbar will remove the files. I looked for occurances of `crowbar-router-migration-failed` and `crowbar-router-migration-ok`, but found none, only this file creates them, then we remove them from crowbar. What is the purpose of these files?

jsuchome replied:
The scripts are executed in background (over ssh) and the caller is checking for presence of these files to realize how did they end.

matelakat commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -322,6 +322,8 @@ def upgrade_first_cluster_node(node, other_node)
         other_node_api = Api::Node.new other_node.name
         node_api.save_node_state("controller")
         save_upgrade_state("Starting the upgrade of node #{node.name}")
+        return false unless evacuate_network_node(node, node["hostname"])
```
Why don't we force namespace cleanup with the first cluster node?

rhafer replied:
All services are still up and running and we don't need the ssh step in that case.

rhafer commented:
> Could you amend the typos, and also remove the references to the removed function from the tests? Also I have some questions inline. I am a bit confused about the role of the -ok -failed files, but I might lack some context.

@matelakat I think I addressed most of the concerns you raised. I believe the timeout thing needs a somewhat broader discussion and the solution should not only over the router_migration step. Are you ok if I create a followup item in the backlog for it? For the namespace deletion issue we just discussed in mumble I created: https://bugzilla.suse.com/show_bug.cgi?id=1018582 as that needs to be addressed in neutron-ha-tool.

matelakat commented:
> @matelakat I think I addressed most of the concerns you raised. I believe the timeout thing needs a somewhat broader discussion and the solution should not only over the router_migration step. Are you ok if I create a followup item in the backlog for it? For the namespace deletion issue we just discussed in mumble I created: https://bugzilla.suse.com/show_bug.cgi?id=1018582 as that needs to be addressed in neutron-ha-tool.

Coo, thanks!
